src/nvim/os/keyboard.c

```diff
@@ -0,0 +1,250 @@
+       || uv_dlsym(&xlib, "XkbGetState", (void **)&_XkbGetState)) != 0) {
+    *err = "Unable to get the functions of X11";
+    return false;
+  }
+  return xlib_opened;
+}
+
+/// Get information about Num/Caps/Scroll Lock state in linux
+///
+/// To be used in os_mods_status() function.
+///
+/// @param[out]  mods  Holds the information about the status of various locks
+///                    in form of a bitmask
+/// @param[out]  err  Location where error message is to be saved, set to NULL
+///                   if no error.
+///
+/// @return true in case of error, false otherwise.
+static bool os_get_locks_status(int *const mods, const char **const err)
+{
+#define CLOSE_DISPLAY(display) \
+  do { \
+    const int e = _XCloseDisplay(display); \
+    if (e) { \
+      *err = "XCloseDisplay unable to close the display"; \
+      return false; \
+    } \
+  } while (0)
+  *mods = 0;
+  *err = NULL;
+  if (!os_xlib_init(err)) {
+    return false;
+  }
+
+  _XDisplay *const display = _XOpenDisplay(NULL);
+  if (display == NULL) {
+    *err = "Couldn't open the display";
+    return false;
+  }
+
+  _XkbStateRec xkb_state;
+  const _XStatus status = _XkbGetState(display, _XkbUseCoreKbd, &xkb_state);
+  if (status) {
+    *err = "Non-zero status returned by XkbGetState";
+    CLOSE_DISPLAY(display);
+  }
+  *mods = xkb_state.data[kLockedModsOffset];
+
+  CLOSE_DISPLAY(display);
+  return true;
+#undef CLOSE_DISPLAY
+}
+
+#elif defined(WIN32)
+/// Get information about Num/Caps/Scroll Lock state in Windows
+///
+/// To be used in os_mods_status() function.
+///
+/// @param[out]  mods  Holds the information about the status of various locks
+///                    in form of a bitmask
+/// @param[out]  err  Location where error message is to be saved, set to NULL
+///                   if no error.
+///
+/// @return true in case of error, false otherwise.
+bool os_get_locks_status(int *const mods, const char **const err)
+{
+  *mods = 0;
+  *err = NULL;
+  if (GetKeyState(VK_CAPITAL) & 0x0001) {
+    *mods = *mods | kCapsLock;
+  }
+  if (GetKeyState(VK_NUMLOCK) & 0x0001) {
+    *mods = *mods | kNumLock;
+  }
+  if (GetKeyState(VK_SCROLL) & 0x0001) {
+    *mods = *mods | kScrollLock;
+  }
+  return true;
+}
+
+#elif defined(__APPLE__)
+/// Get information about Num/Caps Lock state in MacOS
+///
+/// To be used in os_mods_status() function.
+///
+/// @param[out]  mods  Holds the information about the status of various locks
+///                    in form of a bitmask
+/// @param[out]  err  Location where error message is to be saved, set to NULL
+///                   if no error.
+///
+/// @return true in case of error, false otherwise.
+static bool os_get_locks_status(int *const mods, const char **const err)
+{
+  *err = NULL;
+  io_service_t ios;
+  io_connect_t ioc;
+  CFMutableDictionaryRef mdict;
+  bool stateCaps, stateNums;
+  mdict = IOServiceMatching(kIOHIDSystemClass);
+  ios = IOServiceGetMatchingService(kIOMasterPortDefault,
```

Comment:
`ios` should be declared here. Based on the later interactions maybe also be `const io_service_t`.